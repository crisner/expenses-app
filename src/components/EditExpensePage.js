import React from 'react';

const EditExpensePage = () => <div>From the edit page</div>;

export default EditExpensePage;