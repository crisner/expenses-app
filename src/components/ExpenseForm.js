import 'react-dates/initialize';
import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends Component {
    state = {
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        calendarFocused: false
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if(amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = createdAt => {
        this.setState(() => ({ createdAt }));
    };

    onFocusChange = ({focused}) => {
        this.setState(() => ({ calendarFocused: focused }));
    };



    render() {
        return (
            <form>
                <input type="text" placeholder="Description" value={this.state.description} onChange={this.onDescriptionChange} autoFocus />
                <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange} />
                <SingleDatePicker
                 date={this.state.createdAt}
                 onDateChange={this.onDateChange}
                 focused={this.state.focused}
                 onFocusChange={this.onFocusChange}
                 id="datepicker" />
                <textarea placeholder="Enter note(optional)" value={this.state.note} onChange={this.onNoteChange}></textarea>
                <button>Submit</button>
            </form>
        );
    }
}