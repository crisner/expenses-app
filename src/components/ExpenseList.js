import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectedExpenses from '../selectors/expenses';

const ExpenseList = props => (
    <div>
        <h2>Expense List</h2>
        <ul>
            {props.expenses.map(expense => {
                return (
                    <li key={expense.id}>
                        <ExpenseListItem 
                        {...expense} />
                    </li>
                );
            })}
        </ul>
    </div>
);

const mapStateToProps = state => ({ expenses: selectedExpenses(state.expenses, state.filters) });

export default connect(mapStateToProps)(ExpenseList);