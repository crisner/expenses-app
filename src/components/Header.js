import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <div>
        <h2>Expenses</h2>
        <NavLink exact to="/" activeClassName="is-active">Dashboard</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
        <NavLink to="/edit" activeClassName="is-active">Edit Expense</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help</NavLink>
    </div>
);

export default Header;