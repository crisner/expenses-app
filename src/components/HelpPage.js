import React from 'react';

const HelpPage = () => <div>From the help page</div>;

export default HelpPage;