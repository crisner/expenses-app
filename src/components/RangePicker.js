import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setStartDate, setEndDate } from '../actions/filters';
import moment from 'moment';
import Helmet from 'react-helmet';

import DayPickerInput from 'react-day-picker/DayPickerInput';
import 'react-day-picker/lib/style.css';

import { formatDate, parseDate } from 'react-day-picker/moment';

export class RangePicker extends Component {
    state = {
        from: moment(this.props.filters.startDate).toDate(),
        to: moment(this.props.filters.endDate).toDate(),
    };

    showFromMonth = () => {
        const { from, to } = this.state;
        if (!from) {
          return;
        }
        if (moment(to).diff(from, 'months') < 2) {
          this.to.getDayPicker().showMonth(from);
        }
    };
    
    handleFromChange = (from) => {
        // Change the from date and focus the "to" input field
        this.setState({ from });
        this.props.setStartDate(from);
    };

    handleToChange = (to) => {
        this.setState({ to }, this.showFromMonth);
        this.props.setEndDate(to);
    };

    render() {
        const { from, to } = this.state;
        const modifiers = { start: from, end: to };
        
        return (
            <div className="InputFromTo">
                <DayPickerInput
                value={from}
                placeholder="From"
                format="ll"
                formatDate={formatDate}
                parseDate={parseDate}
                dayPickerProps={{
                    selectedDays: [from, { from, to }],
                    disabledDays: { after: to },
                    toMonth: to,
                    modifiers,
                    numberOfMonths: 2,
                    onDayClick: () => this.to.getInput().focus(),
                }}
                onDayChange={this.handleFromChange}
                />{' '}
                â€”{' '}
                <span className="InputFromTo-to">
                <DayPickerInput
                    ref={el => (this.to = el)}
                    value={to}
                    placeholder="To"
                    format="ll"
                    formatDate={formatDate}
                    parseDate={parseDate}
                    dayPickerProps={{
                    selectedDays: [from, { from, to }],
                    disabledDays: { before: from },
                    modifiers,
                    month: from,
                    fromMonth: from,
                    numberOfMonths: 2,
                    }}
                    onDayChange={this.handleToChange}
                />
                </span>
                <Helmet>
                    <style>{`
                    .InputFromTo .DayPicker-Day--selected:not(.DayPicker-Day--start):not(.DayPicker-Day--end):not(.DayPicker-Day--outside) {
                        background-color: #f0f8ff !important;
                        color: #4a90e2;
                    }
                    .InputFromTo .DayPicker-Day {
                        border-radius: 0 !important;
                    }
                    .InputFromTo .DayPicker-Day--start {
                        border-top-left-radius: 50% !important;
                        border-bottom-left-radius: 50% !important;
                    }
                    .InputFromTo .DayPicker-Day--end {
                        border-top-right-radius: 50% !important;
                        border-bottom-right-radius: 50% !important;
                    }
                    `}</style>
                </Helmet>
            </div>
        );
    }
};

const mapStateToProps = state => {
    return {
        filters: state.filters
    };
}

const mapDispatchToProps = dispatch => ({
    setStartDate: (from) => dispatch(setStartDate(from)),
    setEndDate: (to) => dispatch(setEndDate(to))
})

export default connect(mapStateToProps, mapDispatchToProps)(RangePicker);