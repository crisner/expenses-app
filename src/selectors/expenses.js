import moment from 'moment';

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter(expense => {
        const startDateMatch = !startDate || moment(expense.createdAt).isSameOrAfter(startDate);
        const endDateMatch = !endDate || moment(expense.createdAt).isSameOrBefore(endDate);
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b) => {
        if(sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        }

        if(sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    })
};

export default getVisibleExpenses;