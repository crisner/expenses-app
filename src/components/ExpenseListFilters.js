import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate } from '../actions/filters';

import RangePicker from './RangePicker';

export class ExpenseListFilters extends Component {
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortByChange = (e) => {
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
    };

    render () {
        return (
            <div>
                <label>Search:</label>
                <input type="text" value={ this.props.filters.text } onChange={this.onTextChange} />
                <select value={ this.props.filters.sortBy } onChange={this.onSortByChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <RangePicker />
            </div>
        );
    }
}

const mapStateToProps = state => ({
        filters: state.filters
});

const mapDispatchToProps = dispatch => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);