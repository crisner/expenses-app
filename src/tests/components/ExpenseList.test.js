import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('should render the expenses list', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render message when there are no expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});