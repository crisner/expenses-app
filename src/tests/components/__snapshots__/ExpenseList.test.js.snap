// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render message when there are no expenses 1`] = `
<div>
  <h2>
    Expense List
  </h2>
  <p>
    No expenses to show
  </p>
</div>
`;

exports[`should render the expenses list 1`] = `
<div>
  <h2>
    Expense List
  </h2>
  <ul>
    <li
      key="0"
    >
      <ExpenseListItem
        amount={25000}
        createdAt={"1969-12-27T00:00:00.000Z"}
        description="Downpayment"
        id="0"
        note="For renting the new apartment"
      />
    </li>
    <li
      key="1"
    >
      <ExpenseListItem
        amount={6000}
        createdAt={"1970-01-06T00:00:00.000Z"}
        description="Rent"
        id="1"
        note="Starting rent for the month"
      />
    </li>
    <li
      key="2"
    >
      <ExpenseListItem
        amount={200}
        createdAt={"1970-01-01T00:00:00.000Z"}
        description="Phone bill"
        id="2"
        note=""
      />
    </li>
  </ul>
</div>
`;
